apps/emqx/src/emqx_channel.erl

```diff
@@ -1428,19 +1436,17 @@ terminate(_, #channel{conn_state = idle} = _Channel) ->
     ok;
 terminate(normal, Channel) ->
     run_terminate_hook(normal, Channel);
-terminate({shutdown, kicked}, Channel) ->
-    run_terminate_hook(kicked, Channel);
 terminate({shutdown, Reason}, Channel) when
-    Reason =:= discarded;
-    Reason =:= takenover
+    Reason =:= expired orelse
+        Reason =:= takenover orelse
+        Reason =:= kicked orelse
+        Reason =:= discarded
 ->
-    run_terminate_hook(Reason, Channel);
-terminate(Reason, Channel = #channel{clientinfo = ClientInfo, will_msg = WillMsg}) ->
-    %% since will_msg is set to undefined as soon as it is published,
-    %% if will_msg still exists when the session is terminated, it
-    %% must be published immediately.
-    WillMsg =/= undefined andalso publish_will_msg(ClientInfo, WillMsg),
-    run_terminate_hook(Reason, Channel).
+    Channel1 = maybe_publish_will_msg(Reason, Channel),
+    run_terminate_hook(Reason, Channel1);
```

Comment:
Nit. This clause's logic is identical to the next catch-all clause, the only difference is taking the inner atom from `shutdown` tuple. I'd argue this makes the shutdown logic harder to follow.

In general, I suspect that one thing that could help to make things simpler to follow is separating `maybe_publish_will_msg` into a set of 2 functions: one is for _scheduling_ willmsg publishing (e.g. essentially called only when `sock_closed`), another one is for deciding what to do on channel _termination_ (when there's no point to schedule anything). This should also eliminate the need to do something explicitly on `handle_call(kick, ...)`, i.e. [here](https://github.com/emqx/emqx/blob/6f779dc663e0aab9108aa0a85fe250bff333997e/apps/emqx/src/emqx_channel.erl#L1153).

Reply:
> Nit. This clause's logic is identical to the next catch-all clause, the only difference is taking the inner atom from shutdown tuple. I'd argue this makes the shutdown logic harder to follow.

I thought the same and I did remove it until I found `run_terminate_hook` relies on the Reason (kicked) instead of `{shutdown, kicked}`, the hooks are exposed to the users so I think it better to keep it.

Reply:
> In general, I suspect that one thing that could help to make things simpler to follow is separating maybe_publish_will_msg into a set of 2 functions: one is for scheduling willmsg publishing (e.g. essentially called only when sock_closed), another one is for deciding what to do on channel termination (when there's no point to schedule anything). This should also eliminate the need to do something explicitly on handle_call(kick, ...), i.e. [here](https://github.com/emqx/emqx/blob/6f779dc663e0aab9108aa0a85fe250bff333997e/apps/emqx/src/emqx_channel.erl#L1153).

I found scheduling willmsg publishing will not always work when the process terminates before the timer get fired. 
Simple fix would be to just extend the process lifetime but that is costly and it cannot be done in `terminate` without changing the whole call stack.

Reply:
Ah, good catch. However, it also means that terminate hook now will be called with `expired`, instead of `{shutdown, expired}` as before. Although I'm not sure if former behavior is important to preserve here.

Reply:
> I found scheduling willmsg publishing will not always work when the process terminates before the timer get fired.

I mean, that seems the reason why the logic is hard to follow: currently `maybe_publish_will_msg` is called both from sock_closed context (where scheduling timers makes sense) and terminate context (where scheduling timers doesn't make sense). Yet some of its codepaths end up with timers being scheduled anyway, which is confusing: it's very hard to tell in which context they are scheduled. Thus it seems worthwhile to try to separate this into `maybe_schedule_will_msg/1` and `maybe_publish_will_msg_on_terminate/2`, at least on the first look.